Internal: Document - Docs [ED-7059] (#18579)
elementor/elementor: docs/core/base/document.md

# Document

> The term "Document" is used in the only in the code. It refers a single piece of content that created by Elementor, such as a post, page, or a saved template.

Author: ```suggestion
The term "Document" refers to a single piece of content that created by Elementor, such as a post, page, or a saved template. It's used only in the code.
```

> 

Reviewer: I think you should firstly explain what a Document is and why it exists, and then say that it's only a code-term

Each document has its own type and features.

> A document is saved as a `post` in the WordPress `wp_posts` table. His elements and settings are saved as metadata in the `wp_postmeta` table.

Reviewer: ```suggestion
A document is saved as a `post` in WordPress' `wp_posts` table, while its elements and settings are saved as metadata in the `wp_postmeta` table.
```

> A document is saved as a `post` in the WordPress `wp_posts` table. His elements and settings are saved as metadata in the `wp_postmeta` table.

Reviewer: This whole introduction doesn't really feel smooth to read. It feels like a bunch of list items. Maybe try to make it more fluent

In PHP, the document is represented by the `Elementor\Base\Document` class.

> This class is like a DB Model that knows how to save and load its data.

Reviewer: ```suggestion
This class is an abstraction over the `wp_posts` table, that behaves as a DB Model and handles reading and writing the Document data to the database.
```
Maybe it sounds better? IDK

Author: It's not a real abstraction

### Why "Document"?

> WordPress' content are called `posts` it's an historical term that came from the days of WordPress as a blogging system.

Reviewer: ```suggestion
In WordPress, contents are called `posts`. It's an historical term that was introduced in the days when WordPress was just a blogging system instead of the CMS it became to be today.
```

Maybe something like this

> In Elementor, it can be many document types for one `post_type` like the saved templates Page and Section that are both a `elementor_library` post type.

Reviewer: ```suggestion
In Elementor, on the other hand, many document types can share the same `post_type`. For example, "Page" and "Section" templates both have an `elementor_library` post type even though those are different document types.
```

> The term `document` is used instead, as the browser called his DOM content as a "document".

Reviewer: It doesn't really explain why "document" is a  better name than "post", and I don't really understand how it's related to the browser if it's a DB entity

Author: In order to not be confused with WP post.

### Main Features

* Get the data from DB
* Save the data in DB
* Get the elements and settings
* Get metadata like URLs to create, edit and preview the document
* Get the WP_Post object

> * Get the list of widgets categories to display in the panel

Reviewer: It's more like an API documentation for the Document class itself. Maybe just add a list of public methods and their description

### How to create a new Document type?

A new type of document can be used in order to create special functionality for it.

> For example, an `Header` in Elementor Pro is a new type of document that knows to appear in the header of the site.

Reviewer: ```suggestion
For example, a `Header` in Elementor Pro is a new type of document that knows to appear in the header of the site.
```

> A new document type should be created by extending the `Elementor\Base\Document` class, and define his type and the post type it will be used in.

Reviewer: ```suggestion
A new document type should be created by extending the `Elementor\Base\Document` class, and defining its type and the post type it will be used in.
```

> ```php

Reviewer: ```suggestion
#### Example:
```php
```

Maybe?

class Article extends \Elementor\Core\Base\Document {
	public static function get_type() {
		return 'article';

> 	}

Reviewer: Maybe as a better practice for new developers, use a constant for the type and return it herel ike we do in other places.

Author: It should be a simple example

	
	public static function get_properties() {
		$properties = parent::get_properties();
	
		// Which post type the document will be used in
		$properties['cpt'] = [ 'article' ];
	

> 		return $properties;

Reviewer: Not related, but we need a better API for that

	}
}
```

> And then it should be registered the new document type:

Reviewer: ```suggestion
And then, register the new document type:
```

```php

> add_action( 'elementor/documents/register', function( $documents_manager ) {

Reviewer: ```suggestion
add_action( 'elementor/documents/register', function( Documents_Manager $documents_manager ) {
```

> 	$documents_manager->register_document_type( Article::get_type(), Article::get_class_full_name() );

Reviewer: why not like that?
```suggestion
	$documents_manager->register_document_type( Article::get_type(), Article::class );
```

Author: For support autoloading, see https://www.php.net/manual/en/language.oop5.basic.php#:~:text=That%20means%20at%20the%20time%20the%20class%20name%20string%20is%20created%20no%20autoloading%20has%20happened%20yet

} );
```

On registering, the URL to create new Article document is available at:
```php
$create_url = Article::get_create_url();

> ```

Reviewer: It's kinda repeating what you already said about the API methods above. Not sure if it's really necessary

### A Template Document

> In order to create a document type as a template (For internal use) it should extend the `\Elementor\Modules\Library\Documents\Library_Document` class without set the `cpt`, it will use the templates cpt.

Reviewer: Maybe explain what a template is, before telling how to create one

> In order to create a document type as a template (For internal use) it should extend the `\Elementor\Modules\Library\Documents\Library_Document` class without set the `cpt`, it will use the templates cpt.

Reviewer: ```suggestion
In order to create a document type as a template (For internal use) it should extend the `\Elementor\Modules\Library\Documents\Library_Document` class without setting a `cpt` since it will use the template's one.
```

```php
class Article extends \Elementor\Modules\Library\Documents\Library_Document {
	public static function get_type() {
		return 'article';
	}
	
	public static function get_title() {

> 		return __( 'Article', 'plugin-domain' );

Author: ```suggestion
		return __( 'Article', 'plugin-textdomain' );
```

> 		return __( 'Article', 'plugin-textdomain' );

Reviewer: ```suggestion
		return esc_html__( 'Article', 'plugin-textdomain' );
```

	}
}
```

### How to change the widget categories order?

The widgets categories order is defined by the `get_editor_panel_categories` method of the document type.
In order to change the order, the `get_editor_panel_categories` method should be overridden.

```php
class Article extends \Elementor\Modules\Library\Documents\Library_Document {
	public static function get_type() {
		return 'article';
	}
	
	public static function get_title() {

> 		return __( 'Article', 'plugin-domain' );

Author: ```suggestion
		return __( 'Article', 'plugin-textdomain' );
```

	}
	

> 	// Assume that the there are widgets that belongs to a new category called `article-elements`

Reviewer: ```suggestion
	// Assume that there are widgets that belong to a new category called `article-elements`.
```

	protected static function get_editor_panel_categories() {
		$categories = [
			'article-elements' => [
				'title' => __( 'Article', 'plugin-textdomain' ),
				'active' => true,
			],
		];

> 		return $categories + parent::get_editor_panel_categories();

Reviewer: ```suggestion
		// Put the `article-elements` category first.
		return $categories + parent::get_editor_panel_categories();
```

	}
}
```

> Once it's registered, it will appear in the templates list: WP Admin > Templates > Add New.

Reviewer: I think this line should somewhere above, after the registration

* Save the data in DB
* Get the elements and settings
* Get metadata like URLs to create, edit and preview the document
* Get the WP_Post object
* Get the list of widgets categories to display in the panel

### How to create a new Document type?

A new type of document can be used in order to create special functionality for it.

For example, a `Header` in Elementor Pro is a new type of document that knows to appear in the header of the site.

A new document type should be created by extending the `Elementor\Base\Document` class, and defining its type and the post type it will be used in.

#### Example:
```php
class Article extends \Elementor\Core\Base\Document {
	public static function get_type() {
		return 'article';
	}
	
	public static function get_properties() {
		$properties = parent::get_properties();
	
		// Which post type the document will be used in
		$properties['cpt'] = [ 'article' ];
	
		return $properties;
	}
}
```

And then, register the new document type:

```php
add_action( 'elementor/documents/register', function( Documents_Manager $documents_manager ) {
	$documents_manager->register_document_type( Article::get_type(), Article::get_class_full_name() );
} );
```

> On registering, the URL to create new Article document is available at:

Author: ```suggestion
Once it's registered, it will appear in the templates list: WP Admin > Templates > Add New.

Them the URL to create new Article document is available at:
```
